docs/architecture/request_migration.md

```diff
@@ -0,0 +1,114 @@
+
+## Overview
+
+Request migration is implemented through a Migration operator that sits in the LLM processing pipeline between the Backend operator and the service backend. When a worker fails during request processing, the migration system preserves the partial generation state and recreates the request on a new worker to continue from where the previous worker left off.
+
+## Architecture Components
+
+### Migration Operator
+
+The Migration operator is integrated into the LLM processing pipeline between the frontend preprocessing and the actual service backends. This positioning allows it to intercept all communication flows and manage failure scenarios transparently.
+
+Key responsibilities:
+- Intercepts all requests and responses flowing through the pipeline
+- Detects worker failure scenarios through error pattern matching
+- Manages retry logic with configurable migration limits
+- Tracks partial response state for seamless continuation
+
+### Migration Limit Configuration
+
+Each worker can be configured with a migration limit parameter that specifies the maximum number of times a request can be migrated to another worker:
+
+- Default behavior: no migration allowed
+- Can be set independently for different engine types
+- Applicable to LLM worker nodes that perform inference
+- Allows engines to override user-specified limits for compatibility
+
+## Token State Tracking and Request Migration
+
+The core of the migration system is its ability to preserve and continue partial generations through careful token state management. This ensures that when a worker fails mid-generation, the new worker can seamlessly continue from the exact point of failure.
+
+### Token Accumulation Process
+
+When a request is being processed and responses are flowing back from a worker, the migration system tracks every token that has been successfully generated:
+
+1. **Initial Request State**: The system starts with the original preprocessed request containing the initial prompt tokens.
+
+2. **Response Tracking**: As each response arrives from the worker, the migration system extracts the newly generated tokens and appends them to the request's token sequence. This creates an ever-growing accumulation of all tokens that have been successfully generated.
+
+3. **Token Count Management**: The system also updates the remaining token budget to reflect the number of tokens already generated, ensuring that the total generation stays within the originally requested limits.
+
+### Migration Trigger Scenarios
+
+The migration system handles two distinct failure scenarios:
+
+#### 1. New Request Migration (Initial Connection Failure)
+
+**Scenario**: Worker is unreachable when creating the initial connection.
+
+**Error Pattern**: Communication system reports no available workers for the request.
```

Comment:
```suggestion
**Error Pattern**: Communication system reports chosen worker instance is unavailable.
```

I think we need to be clearer here - the internal error that there are no instances available - is really that there are no listeners for this specific worker instance - not that there are no workers available -

Reply:
yes, the selected worker instance is offline when the request is sent.